TINY-7228: Improve RTC documentation and error messages with customer feedback (#1916)
tinymce/tinymce-docs: rtc/encryption.md

---
layout: default
title: Encryption information for the RTC protocol
title_nav: RTC Encryption help
description: Useful information for understanding how encryption is used with RTC

> keywords: rtc encrypt encryption decrypt decryption key rotate signature

Author: Probably not. I haven't written many new pages in these docs.

---

This documentation is in progress. Please contact us with any suggestions you think should be here.

> **Caution**: The advice on this page may not guarantee a secure connection. If data secrecy is important to you please consult a security professional.

## Encryption keys are mandatory

> RTC requires an encryption key so {{site.companyname}} can prove to ourselves (and any interested customers) that the RTC server does not store document content in a format that we are able to read. We are very focussed on privacy and end-to-end encryption for this feature.

Author: This is copy/pasted from an email I wrote to a customer, happy to take suggestions. It's intended to be a one line summary of [Tim's blog post](https://www.tiny.cloud/blog/real-time-collaboration-end-to-end-encryption/).

> RTC requires an encryption key so {{site.companyname}} can prove to ourselves (and any interested customers) that the RTC server does not store document content in a format that we are able to read. We are very focussed on privacy and end-to-end encryption for this feature.

Author: I might make some tweaks for correctness, but this is good. Thanks!

Signing is a one way process where the data is not modified and can still be read by anyone. A signature is transmitted along with the data and can be used to validate the data has not been modified since it was approved by whoever has the signing key. RTC requires an [asymmetric signing algorithm]({{site.baseurl}}/rtc/jwt-authentication/#supportedalgorithms) for JWT identity tokens. Asymmetric signing uses a public/private key pair: signatures are created with the private key and verified using the public key. The public key stored on the Tiny Cloud servers can only be used to verify signatures, not create new ones, so it is quite safe for us to hold or cache without security implications should it be disclosed.

## Choosing an encryption key

> RTC encryption keys can be any unicode string. Our demo application for example uses the simple fixed key `"not a very secret 🔑"`. Even simple keys provide some obfuscation and protection if the key is disclosed; for details on this see [RTC encryption details](#rtcencryptiondetails) below. Simple fixed keys can be a good way to get a test environment up and running but we strongly recommend using secure keys when collaborating on private production data.

Reviewer: - Can we change `unicode` to `utf-8` or appropriate throughout these docs?
- I am strongly against suggesting that people use insecure things, and then adding a "but please change it before you go to prod" because people never change it before they go to prod
- "provide some obfuscation and protection if the key is disclosed" sounds like one of the words should have a negative in it

Author: > I am strongly against suggesting that people use insecure things, and then adding a "but please change it before you go to prod" because people never change it before they go to prod

I generally agree, but in this case doing that will raise the barrier to entry too high. RTC is far more complicated to set up than any feature we've ever released before, so we need some way to ease people into it. We've even considered backing down on the signed JWT requirement for test environments but that's not viable on our cloud services.

Author: Unicode came from the code, and is technically more correct (JavaScript strings are unicode not utf-8). But if we use utf-8 in other TinyMCE docs I'm happy to change it.

I'll see if I can reword the obfuscation piece, it's already up to the third rewrite. Explaining this stuff in simple terms is _hard_.

Author: There are 3 references to `utf-8` in our docs; two in enterprise spelling and one in a config block. There are 8 uses of `unicode` (10 after including RTC).

For now, I'll stick to the more common term we use; perhaps consult Tyler about changing it across our entire documentation site.

I've pushed up a rewrite of this section; I even split off a sub heading "fixed keys are not secure". Hopefully that helps.

---
layout: default
title: Encryption information for the RTC protocol
title_nav: RTC Encryption help
description: Useful information for understanding how encryption is used with RTC
keywords: rtc encrypt decrypt key rotate signature
---

This documentation is in progress. Please contact us with any suggestions you think should be here.

> **Caution**: The advice on this page may not guarantee a secure connection. If data secrecy is important to you please consult a security professional.

## Encryption keys are mandatory

RTC requires an encryption key so {{site.companyname}} can prove to ourselves (and any interested customers) that the RTC server does not store document content in a format that we are able to read. We are very focussed on privacy and end-to-end encryption for this feature.

## The difference between content encryption and JWT signing

> Encryption is a two way process (encryption and decryption) that transforms data into an encrypted form. Modern encryption algorithms produce an output that is indistinguishable from random data without the decryption key. RTC performs symmetric encryption (meaning encryption and decryption use the same key) on all document content, but the key is never sent to the server. As the key only exists on the client endpoints, the term for this is end-to-end encryption.

Reviewer 2: ```suggestion
Encrypted data is indistinguishable from random data without the decryption key. The RTC plugin performs client-side symmetric encryption (using the same key for encryption and decryption) on the editor content. The encryption key is provided to the user and your server (the client endpoints), and is never sent to the {{site.{{site.cloudname}}}} servers; providing end-to-end encryption.
```

Author: again, I'll correct after committing your suggestions - the key isn't provided _to_ the user, technically, or if it is it's provided by the integrating server. Is `{{site.{{site.cloudname}}}}` valid? I've never seen that before 🧐

> Signing is a one way process where the data is not modified and can still be read by anyone. A signature is transmitted along with the data and can be used to validate the data has not been modified since it was approved by whoever has the signing key. RTC requires an [asymmetric signing algorithm]({{site.baseurl}}/rtc/jwt-authentication/#supportedalgorithms) for JWT identity tokens. Asymmetric signing uses a public/private key pair: signatures are created with the private key and verified using the public key. The public key stored on the Tiny Cloud servers can only be used to verify signatures, not create new ones, so it is quite safe for us to hold or cache without security implications should it be disclosed.

Reviewer 2: ```suggestion
JWT signing does not modify the data, and the data can be read by anyone (the data is not encrypted). A JWT signature is transmitted along with the data and is used to validate that the data has not been modified since the signature (JWT) was created. RTC requires an [asymmetric signing algorithm]({{site.baseurl}}/rtc/jwt-authentication/#supportedalgorithms) for JWT identity tokens. Asymmetric signing uses a public/private key pair: signatures are created with the private key and verified using the public key. The public key stored on the {{site.cloudname}} servers can only be used to verify signatures, not create new ones. This allows the {{site.cloudname}} server to verify that the data has not been tampered between the end-points, but does not allow {{site.companyname}} to decrypt the data.
```

Author: I was deliberately trying to describe signing in general, the process is not unique to JWT, but I guess for anyone unfamiliar with it the difference is meaningless.

## Choosing an encryption key

> RTC encryption keys can be any unicode string. {{site.companyname}} strongly recommends using secure keys when collaborating on private production data, but for test environments a simple fixed key is still valid. Our demo application for example uses the simple fixed key `"not a very secret 🔑"`.

Author: The demo application hasn't been released yet, so I can't link to it. This will do for now.

### Fixed keys are not secure

> Even simple keys provide some protection against data disclosure; for details on this see [RTC encryption details](#rtcencryptiondetails) below. Simple fixed keys can be a good way to get started with RTC but if a key is disclosed it can be used to access any document encrypted with that key. Fixed keys mean _every_ document can be decrypted if the key is disclosed.

Reviewer 2: ```suggestion
Simple, fixed encryption keys provide minimal protection against data disclosure; for details see: [RTC encryption details](#rtcencryptiondetails). Simple fixed keys can be a good way to get started with RTC but if a key is disclosed, it can be used to access any document encrypted with that key. Fixed keys allow _every_ document can be decrypted if the key is disclosed.
```

## Generating a secure encryption key

> Encryption security is a trade off between the complexity of generating a key and the risk of compromise should the key be disclosed to an unknown third party. Here are some suggested ways to generate keys, in descending order of safety:

Reviewer 2: ```suggestion
Encryption security is a trade off between: the complexity of generating a key, and the risk of compromise should the key be disclosed to an unknown third-party. Methods generate keys include (in descending order of safety):
```

In order to allow for key rotation a key hint is supplied so the encryption provider may tell the difference between these two cases and act accordingly. If the key hint is `null`, then the RTC protocol needs a new key and this is the opportunity for key rotation. If the key hint is not null the protocol is requesting a previously issued key so that it can read the session history.

If the key hint is not included in an encryption provider response for a new key, an empty string will be recorded as the key hint.

> **Warning**: The key hint is stored by the server in plain text. Do not add secret or sensitive information to the key hint.

The key hint can be a key thumbprint, ID, or other non-sensitive identifier that will help select the key, such as a timestamp. It is only recorded when the protocol requests a new key. Returning a key hint from requests for previously issued keys will have no effect.

## RTC encryption details

> **Note**: This section contains the technical details of encryption used to securely transmit document contents. It is provided for information purposes only; understanding these details is not required to integrate RTC.

> The RTC protocol does not use the [provided encryption key]({{site.baseurl}}/rtc/configuration#rtc_encryption_provider) to encrypt content directly. Using industry standard cryptography algorithms a unique session key for content encryption is derived from the provided key in a way that makes brute force decryption very difficult.

Reviewer 2: ```suggestion
The RTC plugin does not use the [provided encryption key]({{site.baseurl}}/rtc/configuration#rtc_encryption_provider) to encrypt content directly. The plugin uses industry standard cryptography algorithms to generate a unique session key for content encryption based on the provided key to improve protection against brute force decryption.
```

Author: I appreciate that you're trying to make things easier to follow, but for the target audience here the word "derived" is _very_ important to include. I'll put it back after the suggestions are committed.

> **Note**: {{site.companyname}} is in the process of changing RTC encryption to [AES-KW](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/wrapKey) which does not require an initialization vector or protocol salt. The description below is still correct for the current beta release.

The RTC protocol encryption technique is as follows:

> * 256 bits of random data are generated as the salt for each session. As described above, each document ID used for collaboration may have multiple sessions. The salt is generated with the [Crypto.getRandomValues()](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues) browser API and a `Uint8Array` of length 32 bytes.

Reviewer 2: ```suggestion
* 256 bits of random data are generated as the salt for each session. As described above, each document ID used for collaboration may have multiple sessions. The salt is generated using the [`Crypto.getRandomValues()`](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues) browser API and a 32 byte long [`Uint8Array`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array).
```

> * The salt data and the provided encryption key are combined using the [PBKDF2](https://en.wikipedia.org/wiki/PBKDF2) key derivation function. Derivation is performed by the [SubtleCrypto.deriveKey()](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/deriveKey) browser API, using the following parameters:

Reviewer 2: Regarding the term "key derivation function" and "derivation", we need to either:
- link to a reliable source
- provide a description
- avoid using
I know this is a technical section, but that also means it's easy to get lost.

Author: wikipedia isn't reliable?

Perhaps this entire section needs to be shifted to a dedicated "security professionals, here is how we can prove to you we know what we're talking about" page instead of tacking it onto the end of the general integrator-targeted encryption page.

Reviewer 2: > wikipedia isn't reliable?

no. it's not.

Regarding your second point, maybe? This is something we could try address now or before the RTC 1.0 release.

Author: I could link to the IETF RFC, but that's _completely_ unintelligible. Wikipedia at least gives an average user a chance if they try to follow this.

I think moving it to a separate page can be a pre-1.0 task, rather than trying to figure it out now.

Author: Google for PBKDF2 also brings up https://cryptobook.nakov.com/mac-and-key-derivation/pbkdf2 but that's someone's personal site by the looks of it. A professional someone, but I don't know if it will meet your standard for reliability.

Author: I went with the IETF. Google will bring up wikipedia, we don't need to link to it.

>   * PBKDF2 [algorithm](https://developer.mozilla.org/en-US/docs/Web/API/Pbkdf2Params), with SHA-512 digest algorithm and 10000 iterations

Reviewer 2: ```suggestion
  * PBKDF2 [algorithm](https://developer.mozilla.org/en-US/docs/Web/API/Pbkdf2Params), with SHA-512 digest algorithm and 10000 iterations,
```

>   * AES-GCM [derivedKeyAlgorithm](https://developer.mozilla.org/en-US/docs/Web/API/AesKeyGenParams) with length 256

Reviewer 2: ```suggestion
  * AES-GCM [derivedKeyAlgorithm](https://developer.mozilla.org/en-US/docs/Web/API/AesKeyGenParams) with a length of 256.
```

> * The derived key is cached (in memory only) and used with [SubtleCrypto.encrypt()](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/encrypt) and [SubtleCrypto.decrypt()](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/decrypt) as necessary, specifying the [AES-GCM algorithm](https://developer.mozilla.org/en-US/docs/Web/API/AesGcmParams). For each encrypted message 96 bits of random data are used as the initialization vector, also generated with [Crypto.getRandomValues()](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues) but this time using `Uint8Array` length 12. Each IV is transmitted alongside the encrypted message.

Reviewer 2: ```suggestion
* The derived key is cached (in memory only) and used with [SubtleCrypto.encrypt()](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/encrypt) and [SubtleCrypto.decrypt()](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/decrypt) as necessary, specifying the [AES-GCM algorithm](https://developer.mozilla.org/en-US/docs/Web/API/AesGcmParams). For each encrypted message, 96 bits of random data are used as the initialization vector, also generated with [Crypto.getRandomValues()](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues) but this time using `Uint8Array` with a length of 12. Each IV is transmitted alongside the encrypted message.
```

Reviewer 2: > Each IV is transmitted alongside the encrypted message.

What's `IV`?

Author: Initialization Vector. I guess I can repeat that rather than using the acronym.

> * The provided key and derived key are never transmitted between clients. Each client will already have the provided key, so only the salt needs to be transmitted for clients to recreate the derived key. The salt alone cannot be used to decrypt the document contents.

Reviewer 2: ```suggestion
* The provided key and derived key are never transmitted between clients. Each client will already have the provided key, so only the salt needs to be transmitted for clients to recreate the derived key. The salt alone cannot be used to decrypt the document content.
```

> * When a session is established the provided `keyHint` and protocol-generated salt are sent to the server via a HTTPS websocket connection where they are stored in a database entry linked to the session. When subsequent clients connect to the session, the server sends these values along with other session information so the client can replicate the unique session key using [rtc_encryption_provider]({{site.baseurl}}/rtc/configuration#rtc_encryption_provider) and key derivation.

Reviewer 2: ```suggestion
* When a Real-time collaboration session is established, the provided `keyHint` and protocol-generated salt are sent to the server through a HTTPS websocket connection, where they are stored in a database-entry linked to the session. When subsequent clients connect to the session, the server sends these values along with other session information so the client can replicate the unique session key using [rtc_encryption_provider]({{site.baseurl}}/rtc/configuration#rtc_encryption_provider) and key derivation.
```

Author: `third-party` doesn't have a hyphen, but `database-entry` does? 🧐

Reviewer 2: no, it does. if I missed it, please fix third-party refs (hyphen required)
Added the hyphen here for readability. those three words in a row are hard to parse, so the hyphen was added to show association between to the two most-related words.

Author: Ah my apologies I misremembered the earlier change - you _added_ one, you didn't remove one. There's a lot to keep track of.

> Our choice of encryption parameters results in an expensive derivation process even on 2021 hardware, but the derivation is only performed once by each client when connecting to a session. {{site.companyname}} is committed to security and privacy of customer data; we chose to favour increased security over derivation performance. Once derivation is complete, encryption and decryption will be hardware accelerated.

Reviewer 2: ```suggestion
Our choice of encryption parameters results in an expensive derivation process, even on modern hardware, but the derivation is only performed once by each client when connecting to a session. {{site.companyname}} is committed to the security and privacy of customer data, and {{site.companyname}} has prioritized increased security over derivation performance. Once derivation is complete, encryption and decryption will be hardware accelerated.
```

---
layout: default
title: Understanding encryption for RTC
title_nav: RTC Encryption help
description: Useful information for understanding how encryption is used with RTC
keywords: rtc encrypt decrypt key rotate signature
---

This documentation is in progress. Please contact us with any suggestions you think should be here.

> **Caution**: The advice on this page does not guarantee a secure connection. If data secrecy is important for your users, please consult a security professional.

## Encryption keys are mandatory

{{site.companyname}} is committed to protecting customer and end-user privacy through end-to-end encryption.

{{site.productname}} Real-Time Collaboration requires an encryption key:

- To encrypt content before a client sends it to the {{site.cloudname}} server.
- To decrypt content clients receive from the {{site.cloudname}} server.

{{site.productname}} clients use the encryption key but do not transmit it to the {{site.cloudname}} servers. {{site.companyname}} **will never have the key to decrypt document content**.

## The difference between content encryption and JWT signing

> Encrypted data is indistinguishable from random data without the decryption key. The RTC plugin performs client-side symmetric encryption (using the same key for encryption and decryption) on the editor content. The encryption key is provided to the client by your server, and is never sent to the {{site.cloudname}} servers; that is end-to-end encryption.

Reviewer 2: ```suggestion
Encrypted data is indistinguishable from random data without the decryption key. The RTC plugin performs client-side symmetric encryption (using the same key for encryption and decryption) on the editor content. The encryption key is provided to the client by your server, and is never sent to the {{site.cloudname}} servers; providing end-to-end encryption.
```

> **Note**: {{site.companyname}} is in the process of changing the RTC encryption method to [AES-KW](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/wrapKey), which does not require an initialization vector or protocol salt. The description below is still correct for the current beta release.

The RTC protocol encryption technique is as follows:
* As described above, each document ID used for collaboration may have multiple sessions. To ensure each session has a unique key, 256 bits of random data are generated as the salt for each session. The salt is generated using the [`Crypto.getRandomValues()`](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues) browser API and a 32 byte [`Uint8Array`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array).
* The salt data and the provided encryption key are combined using a [PBKDF2](https://tools.ietf.org/html/rfc2898#section-5.2) key derivation function. Derivation is performed by the [SubtleCrypto.deriveKey()](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/deriveKey) browser API, using the following parameters:

>   * PBKDF2 [algorithm](https://developer.mozilla.org/en-US/docs/Web/API/Pbkdf2Params), with SHA-512 digest algorithm and 10000 iterations,

Reviewer 2: ```suggestion
  * PBKDF2 [algorithm](https://developer.mozilla.org/en-US/docs/Web/API/Pbkdf2Params), with SHA-512 digest algorithm and 10,000 iterations,
```
